linera-execution/src/wasm/async_boundary.rs

```diff
@@ -5,156 +5,240 @@
-    ///
-    /// The error will be returned when this [`GuestFuture`] is polled.
-    FailedToCreate(Option<Application::Error>),
+/// A wrapper type representing the response sent from a future actor to a poll request sent from a
+/// [`PollSender`].
+///
+/// This helps to simplify some types used, avoiding some Clippy lints.
+struct PollResponse<Output>(Poll<Result<Output, ExecutionError>>);
+
+/// An abstraction over a [`Future`] running as an actor on a non-asynchronous thread.
+///
+/// This type implements [`Future`] and sends poll requests to the actor implementation. When the
+/// actor finishes, it sends back the result to this type, which then returns it.
+///
+/// Poll requests may not be sent to the implementation if it would cause non-deterministic
+/// execution (as controlled by the [`HostFutureQueue`]).
+pub struct PollSender<Output> {
+    host_future_queue: Option<HostFutureQueue>,
+    poll_requester: std::sync::mpsc::Sender<oneshot::Sender<PollResponse<Output>>>,
+    state: PollSenderState<Output>,
+}
+
+/// The internal state of the [`PollSender`] type.
+#[derive(Debug)]
+enum PollSenderState<Output> {
+    /// Waiting to be polled.
+    Queued,
 
-    /// The Wasm future type and the runtime context to poll it.
-    Active {
-        /// A WIT resource type implementing a [`GuestFutureInterface`] so that it can be polled.
-        future: Future,
+    /// Waiting for response to the previous poll request.
+    Polling(oneshot::Receiver<PollResponse<Output>>),
 
-        /// Types necessary to call the guest Wasm module in order to poll the future.
-        context: WasmRuntimeContext<'context, Application>,
-    },
+    /// Result received and returned.
+    Finished,
 }
 
-impl<'context, Future, Application> GuestFuture<'context, Future, Application>
-where
-    Application: ApplicationRuntimeContext,
-{
-    /// Creates a [`GuestFuture`] instance with `creation_result` of a future resource type.
+impl<Output> PollSender<Output> {
+    /// Creates a new [`PollSender`] using the provided [`HostFutureQueue`] to ensure deterministic
+    /// polling.
     ///
-    /// If the guest resource type could not be created by the Wasm module, the error is stored so
-    /// that it can be returned when the [`GuestFuture`] is polled.
-    pub fn new(
-        creation_result: Result<Future, Application::Error>,
-        context: WasmRuntimeContext<'context, Application>,
-    ) -> Self {
-        match creation_result {
-            Ok(future) => GuestFuture::Active { future, context },
-            Err(error) => GuestFuture::FailedToCreate(Some(error)),
-        }
+    /// Returns the new [`PollSender`] together with the receiver endpoint of the poll requests.
+    fn new(
+        host_future_queue: Option<HostFutureQueue>,
+    ) -> (
+        Self,
+        std::sync::mpsc::Receiver<oneshot::Sender<PollResponse<Output>>>,
+    ) {
+        let (poll_sender, poll_receiver) = std::sync::mpsc::channel();
```

Comment:
Same as before, I would create the `mpsc` channel in the caller (there's only one place afaict) so that this constructor returns `Self`.

Reply:
Done.